File: solutions/haskell/game-of-life/3/src/GameOfLife.hs
module GameOfLife (tick) where

tick :: [[Int]] -> [[Int]]
tick board =
  [ [cell r c | c <- [0 .. length (board !! r) - 1]]
  | r <- [0 .. length board - 1]
  ]
 where
  cell r c
    | alive && (n == 2 || n == 3) = 1 -- Any live cell with two or three live neighbors lives on.
    | dead && (n == 3) = 1 -- Any dead cell with exactly three live neighbors becomes a live cell.
    | otherwise = 0 -- All other cells die or stay dead.
   where
    alive = ((board !! r) !! c) == 1
    dead = ((board !! r) !! c) == 0
    n = count r c

  count r c =
    let dim = length board - 1
     in sum
          [ board !! y !! x
          | dy <- [-1 .. 1]
          , dx <- [-1 .. 1]
          , (dx, dy) /= (0, 0)
          , let (x, y) = (c + dx, r + dy)
          , y `elem` [0 .. dim]
          , x `elem` [0 .. dim]
          ]

File: solutions/clojure/game-of-life/1/src/game_of_life.clj
(ns game-of-life)

(defn- count-neighbors [cells x y]
  (reduce + (for [dx [-1 0 1]
                  dy [-1 0 1]
                  :when (not= 0 dx dy)]
              (get-in cells [(+ y dy) (+ x dx)] 0))))

(defn tick [cells]
  (map-indexed
   (fn [y row]
     (map-indexed
      (fn [x cell]
        (let [n (count-neighbors cells x y)]
          (if (or (= n 3)
                  (= [cell n] [1 2]))
            1
            0)))
      row))
   cells))
